### PR TITLE
feat: pipeline spec as a separate go module

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
Main arguments for moving pipeline spec to a separate go module:
* it's best if it can have minimal dependencies, because
   * we will need to import pipeline spec from both v1 and v2 code base
   * other tools built on top of KFP may want to import the pipeline spec alone

Scope, this PR will only generate the pipeline spec. I'll migrate all existing usages in a separate PR if we agree with the idea.

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by Bobgy** [comment]
cc @capri-xiyue @ajaygopinathan

File: api/Makefile [hunk 1]
```diff
@@ -1,13 +1,22 @@
...
 		pipeline_spec.proto
 
-gen_cache_key_proto: v2alpha1/cache_key.proto
-	cd v2alpha1 && protoc -I=$(PROTO_PATH) -I=. \
-		--go_out=go \
+go_cachekey: v2alpha1/pipeline_spec.proto v2alpha1/cache_key.proto
+	go install google.golang.org/protobuf/cmd/protoc-gen-go
+	mkdir -p v2alpha1/go/cachekey
+	cd v2alpha1 && protoc -I=. \
+		--go_out=go/cachekey \
```
Comment (Bobgy) [inline api/Makefile:20]: note, cache_key is used internally, so it's put to a different folder from pipeline spec.

File: api/v2alpha1/go/pipelinespec/pipeline_spec.pb.go [hunk 2]
```diff
@@ -1,17 +1,16 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.25.0
-// 	protoc        v3.17.3
+// 	protoc-gen-go v1.27.1
+// 	protoc        v3.6.1
```
Comment (Bobgy) [inline api/v2alpha1/go/pipelinespec/pipeline_spec.pb.go:4]: @capri-xiyue why did you need a new version of protoc?

I installed protoc from `sudo apt install protobuf-compiler`, it's a little old, but convenient to install. Can I keep using the old version?

Comment (capri-xiyue) [reply api/v2alpha1/go/pipelinespec/pipeline_spec.pb.go:4]: If the old version support protocol reflect, it should be ok. I use new versions because I need the protocol reflect method so that I can use protocol json to do marshal.

**Comment by capri-xiyue** [comment]
This PR looks good to me

File: api/v2alpha1/go/pipelinespec/pipeline_spec.pb.go (hunk 2 above)
Comment (Bobgy) [reply api/v2alpha1/go/pipelinespec/pipeline_spec.pb.go:4]: Rethinking over this, I think we should use gcr.io/ml-pipeline-test/api-generator directly, so that we can ensure all versions match despite developer.
